### PR TITLE
R2 final review

### PR DESCRIPTION
**Pull request by ServiceDog**
NOTE: There are a lot of unwraps() in this release. Since we're exploring moving the router to a PSQL or Redis database, I would prefer to not address these unwraps in this release as those blocks may disappear altogether in the next release.

### CODE REVIEW

**Comment by GoodluckH** [comment]
## Documents
- [ ] ICD
  - [ ] Current interfaces (grpc & rest)
  - [ ] Current relevant document links
- [ ] SDD
  - [ ] Description of grpc & rest handler implementations
  - [ ] Current sequence diagrams where applicable
  - [ ] Current relevant document links
- [ ] CONOPS
  - [ ] Current information
  - [ ] Bibliography (if any citations)
- [ ] README (see svc-cargo for example)
  - [ ] Badges
  - [ ] Accurate installation instructions
  - [ ] Current links to other documents
  - [ ] Arrow DAO links
- [ ] Any instance of :construction: should genuinely be under construction

## Code (not applicable to examples/ folder)
- [ ] ~No unhandled unwrap()s~ N/A
- [ ] Accurate (!) code comments on all functions
- [ ] No `# Arguments` and `# Returns` in function comments
- [ ] Function arguments checked where possible
- [ ] Sufficient logger (info, warn, and debug) messages
- [ ] Sensible Types (unsigned vs. signed)
- [ ] No `#[allow(dead_code)]` without good reason
- [ ] End all error, info, and debug messages with punctuation (periods, etc.)
- [ ] No FIXMEs or TODOs without a projected release (`TODO R4: etc.`)

## REST API
- [ ] Return codes accurately documented with openapi
- [ ] Return codes make sense
- [ ] Sufficiently descriptive error messages
- [ ] Do NOT reflect a user's bad input back at them in replies.
    - i.e. "Warning: '{}' is not a valid entry." should instead be "Warning: invalid entry."
    - Use debug messages to see the invalid argument.

## Cargo.toml(s)
- [ ] Package versions correct
- [ ] No minor versions specified
- [ ] Links to crates.io package, not git branch
- [ ] Must have `license-file = "LICENSE"` under `[package]`

**Review by GoodluckH — CHANGES_REQUESTED** [review]
Now that the `lib-router` has been combined with scheduler, please update CONOPS with this new information in mind.

File: README.md [hunk 1]
```diff
@@ -2,12 +2,21 @@
 
 # `svc-scheduler`
 
+![GitHub release (latest by date including pre-releases)](https://img.shields.io/github/v/release/Arrow-air/svc-scheduler?include_prereleases)
+![Sanity Checks](https://github.com/arrow-air/svc-scheduler/actions/workflows/sanity_checks.yml/badge.svg?branch=main)
 ![Rust
 Checks](https://github.com/arrow-air/svc-scheduler/actions/workflows/rust_ci.yml/badge.svg?branch=main)
 ![Python Flake8](https://github.com/arrow-air/svc-scheduler/actions/workflows/python_ci.yml/badge.svg?branch=main)
```
Comment (GoodluckH) [inline README.md:9]: Change it to Pep8

File: server/src/router/router_utils/generator.rs [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+const RAD_TO_DEG: f32 = 180.0 / std::f32::consts::PI;
+
+#[allow(dead_code)]
+/// Generate a vector of random nodes.
+///
+/// # Arguments
+/// * `capacity` - The number of nodes to generate.
+///
+/// # Returns
+/// A vector of nodes.
```
Comment (GoodluckH) [inline server/src/router/router_utils/generator.rs:25]: remove

File: server/src/router/router_utils/generator.rs [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+
+/// Generate a vector of random nodes near a location.
+///
+/// # Arguments
+/// * `location` - The location to generate nodes near.
+/// * `radius` - The radius in kilometers to generate nodes within.
+/// * `capacity` - The number of nodes to generate.
+///
+/// # Returns
+/// A vector of nodes.
```
Comment (GoodluckH) [inline server/src/router/router_utils/generator.rs:50]: remove

File: server/src/router/router_utils/generator.rs [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+}
+
+/// Generate a random node near a location within radius in kilometers.
+///
+/// # Arguments
+/// * `location` - The location to generate nodes near.
+/// * `radius` - The radius in kilometers to generate nodes within.
+///
+/// # Returns
+/// A node with a location near the given location.
```
Comment (GoodluckH) [inline server/src/router/router_utils/generator.rs:90]: remove

File: server/src/router/router_utils/generator.rs [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+}
+
+/// Generate a random location near a given location and radius.
+///
+/// # Arguments
+/// * `location` - The location to generate a random location near.
+/// * `radius` - The radius in kilometers.
+///
+/// # Returns
+/// A random location near the given location and radius.
```
Comment (GoodluckH) [inline server/src/router/router_utils/generator.rs:128]: remove

File: server/src/router/router_utils/generator.rs [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+/// Source: [Reddit](https://www.reddit.com/r/rust/comments/f08lqu/comment/fgsxeik/)
+///
+/// # Arguments
+/// * `rng` - The random number generator.
+/// * `latitude` - The latitude of the location.
+/// * `longitude` - The longitude of the location.
+/// * `radius` - The radius in kilometers.
+///
+/// # Returns
+/// A latitude and longitude pair.
```
Comment (GoodluckH) [inline server/src/router/router_utils/generator.rs:157]: remove

File: server/src/router/router_utils/graph.rs [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+/// # Arguments
+/// * `nodes` - A vector of nodes.
+/// * `constraint` - Only nodes within a constraint can be connected.
+/// * `constraint_function` - A function that takes two nodes and
+///   returns a float to compare against `constraint`.
+/// * `cost_function` - A function that computes the "weight" between
+///   two nodes.
+///
+/// # Returns
+/// A vector of edges in the format of (from_node, to_node, weight).
```
Comment (GoodluckH) [inline server/src/router/router_utils/graph.rs:27]: remove

File: server/src/router/router_utils/haversine.rs [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+use crate::router::router_types::location::Location;
+
+/// Calculate the distance between two points on a sphere.
+///
+/// # Arguments
+/// * `from` - The starting point.
+/// * `to` - The ending point.
+///
+/// # Returns
+/// The distance between the two points in kilometers.
```
Comment (GoodluckH) [inline server/src/router/router_utils/haversine.rs:18]: remove

**Review by GoodluckH — COMMENTED** [review]
Some TODOs from the `lib-router` doesn't have to have a release associated with them. It more like "lazy" TODOs where we address them when necessary.

File: server/src/router/router_utils/generator.rs (hunk 6 above)
Comment (GoodluckH) [inline server/src/router/router_utils/generator.rs:157]:
```suggestion
```

File: server/src/router/router_utils/graph.rs (hunk 7 above)
Comment (GoodluckH) [reply server/src/router/router_utils/graph.rs:27]: actually, i feel like this doc comment is helpful.

File: server/src/router/router_utils/haversine.rs (hunk 8 above)
Comment (GoodluckH) [inline server/src/router/router_utils/haversine.rs:18]:
```suggestion
```

File: server/src/router/router_utils/generator.rs (hunk 2 above)
Comment (GoodluckH) [inline server/src/router/router_utils/generator.rs:25]:
```suggestion
```

File: server/src/router/router_utils/generator.rs (hunk 3 above)
Comment (GoodluckH) [inline server/src/router/router_utils/generator.rs:50]:
```suggestion
```

File: server/src/router/router_utils/generator.rs (hunk 4 above)
Comment (GoodluckH) [inline server/src/router/router_utils/generator.rs:90]:
```suggestion
```

File: README.md (hunk 1 above)
Comment (ServiceDog) [reply README.md]: https://github.com/Arrow-air/svc-scheduler/pull/36/commits/cd81b978e79180ca872b794a32e7fb03f7ff3663

**Comment by ServiceDog** [comment]
> Some TODOs from the `lib-router` doesn't have to have a release associated with them. It more like "lazy" TODOs where we address them when necessary.

Added R3 to all remaining TODOs

**Review by GoodluckH — APPROVED** [review]

**Comment by GoodluckH** [comment]
await @owlot for final approval.

**Review by owlot — APPROVED** [review]

**Pull request merged**